From ee069e38d148c88c19859ef8c962c53d9b1a13a2 Mon Sep 17 00:00:00 2001
From: Sveinung Kvilhaugsvik <sveinung84@users.sourceforge.net>
Date: Thu, 21 Jul 2016 19:38:42 +0200
Subject: [PATCH 3/3] savegame2: tolerate a base becoming a pure extra.

Some extras doesn't naturally fit in as bases. Before 2.6 they still had to
be implemented as bases. It should therefore be expected that some rulesets
will drop the base part of an extra. 3.0 has already done this for Ruins.

Relax the restriction that the number of bases in a save file must be lesser
than or equal to the number of bases in the ruleset by giving an exception
to the rule if all bases in the save file are defined as extras.

See patch #7493
---
 server/savegame2.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/server/savegame2.c b/server/savegame2.c
index dfd2456..bd82c12 100644
--- a/server/savegame2.c
+++ b/server/savegame2.c
@@ -1272,16 +1272,17 @@ static void sg_load_savefile(struct loaddata *loading)
     sg_failure_ret(loading->base.size != 0,
                    "Failed to load bases order: %s",
                    secfile_error());
-    sg_failure_ret(!(game.control.num_base_types < loading->base.size),
-                   "Number of bases defined by the ruleset (= %d) are "
-                   "lower than the number in the savefile (= %d).",
-                   game.control.num_base_types, (int)loading->base.size);
     /* make sure that the size of the array is divisible by 4 */
     nmod = 4 * ((loading->base.size + 3) / 4);
     loading->base.order = fc_calloc(nmod, sizeof(*loading->base.order));
     for (j = 0; j < loading->base.size; j++) {
       struct extra_type *pextra = extra_type_by_rule_name(modname[j]);
 
+      sg_failure_ret(pextra != NULL
+                     || game.control.num_base_types >= loading->base.size,
+                     "Unknown base type %s in savefile.",
+                     modname[j]);
+
       if (pextra != NULL) {
         loading->base.order[j] = extra_base_get(pextra);
       } else {
-- 
2.1.4

